Elasticsearch doesn't work in chroots: requires / to be a mountpoint

For somewhat arbitrary reason, Elasticsearch's file locking code does a bunch of sanity checks and will error out if `/` is not a mountpoint. This is not the case when running in a chroot.

While the correct fix is definitely for Elasticsearch to accept to run in chroots (elastic/elasticsearch#12018), there might be some possible workarounds in reprounzip-vagrant.

* It is possible to make `/` a mountpoint by bind-mounting it from somewhere else and chrooting in the mount
* Also, might be good to create a fake fstab? With at least an entry for `/`? Docker mounts a few key files into the container (such as hosts, resolv.conf), we might want to do that

still not work in chroot.

Caused by: java.io.IOException: Mount point not found
        at sun.nio.fs.LinuxFileStore.findMountEntry(LinuxFileStore.java:91) ~[?:?]
        at sun.nio.fs.UnixFileStore.<init>(UnixFileStore.java:65) ~[?:?]
        at sun.nio.fs.LinuxFileStore.<init>(LinuxFileStore.java:44) ~[?:?]
        at sun.nio.fs.LinuxFileSystemProvider.getFileStore(LinuxFileSystemProvider.java:51) ~[?:?]
        at sun.nio.fs.LinuxFileSystemProvider.getFileStore(LinuxFileSystemProvider.java:39) ~[?:?]
        at sun.nio.fs.UnixFileSystemProvider.getFileStore(UnixFileSystemProvider.java:368) ~[?:?]
        at java.nio.file.Files.getFileStore(Files.java:1461) ~[?:1.8.0_192]
        at org.elasticsearch.env.Environment.getFileStore(Environment.java:324) ~[elasticsearch-6.5.4.jar:6.5.4]
        at org.elasticsearch.env.NodeEnvironment$NodePath.<init>(NodeEnvironment.java:108) ~[elasticsearch-6.5.4.jar:6.5.4]
        at org.elasticsearch.env.NodeEnvironment$NodeLock.<init>(NodeEnvironment.java:209) ~[elasticsearch-6.5.4.jar:6.5.4]
        at org.elasticsearch.env.NodeEnvironment.<init>(NodeEnvironment.java:267) ~[elasticsearch-6.5.4.jar:6.5.4]
        at org.elasticsearch.node.Node.<init>(Node.java:296) ~[elasticsearch-6.5.4.jar:6.5.4]
        at org.elasticsearch.node.Node.<init>(Node.java:265) ~[elasticsearch-6.5.4.jar:6.5.4]
        at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:212) ~[elasticsearch-6.5.4.jar:6.5.4]
        at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:212) ~[elasticsearch-6.5.4.jar:6.5.4]
        at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:333) ~[elasticsearch-6.5.4.jar:6.5.4]
        at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:136) ~[elasticsearch-6.5.4.jar:6.5.4]

--- a/reprounzip-vagrant/reprounzip/unpackers/vagrant/__init__.py
+++ b/reprounzip-vagrant/reprounzip/unpackers/vagrant/__init__.py
@@ -171,13 +171,13 @@ def machine_setup(target):
         chan = ssh.get_transport().open_session()
         chan.exec_command(
             '/usr/bin/sudo /bin/sh -c %s' % shell_escape(
-                'if ! grep -q "/experimentroot " /proc/mounts; then '
+                'if ! grep -q "/experimentroot " /etc/mtab; then '
                 'mount -o bind /.experimentdata /experimentroot; '
                 'fi; '
-                'if ! grep -q "/experimentroot/dev " /proc/mounts; then '
+                'if ! grep -q "/experimentroot/dev " /etc/mtab; then '
                 'mount -o rbind /dev /experimentroot/dev; '
                 'fi; '
-                'if ! grep -q "/experimentroot/proc " /proc/mounts; then '
+                'if ! grep -q "/experimentroot/proc " /etc/mtab; then '
                 'mount -t proc none /experimentroot/proc; '
                 'fi'))
         if chan.recv_exit_status() != 0:

--- a/reprounzip-vagrant/reprounzip/unpackers/vagrant/__init__.py
+++ b/reprounzip-vagrant/reprounzip/unpackers/vagrant/__init__.py
@@ -164,6 +164,7 @@ def machine_setup(target):
 
     if use_chroot:
         # Mount directories
+        logger.debug("Mounting directories")
         ssh = paramiko.SSHClient()
         ssh.set_missing_host_key_policy(IgnoreMissingKey())
         ssh.connect(**info)
@@ -184,6 +185,7 @@ def machine_setup(target):
             sys.exit(1)
         if gui:
             # Mount X11 socket
+            logger.debug("Mounting X11 socket")
             chan = ssh.get_transport().open_session()
             chan.exec_command(
                 '/usr/bin/sudo /bin/sh -c %s' % shell_escape(
@@ -197,6 +199,8 @@ def machine_setup(target):
                 logger.critical("Couldn't mount X11 sockets in chroot")
                 sys.exit(1)
         ssh.close()
+    else:
+        logger.debug("NOT mounting directories")
 
     return info
 

--- a/reprounzip-vagrant/reprounzip/unpackers/vagrant/__init__.py
+++ b/reprounzip-vagrant/reprounzip/unpackers/vagrant/__init__.py
@@ -170,6 +170,9 @@ def machine_setup(target):
         chan = ssh.get_transport().open_session()
         chan.exec_command(
             '/usr/bin/sudo /bin/sh -c %s' % shell_escape(
+                'if ! grep -q "/experimentroot " /proc/mounts; then '
+                'mount -o bind /.experimentdata /experimentroot; '
+                'fi; '
                 'if ! grep -q "/experimentroot/dev " /proc/mounts; then '
                 'mount -o rbind /dev /experimentroot/dev; '
                 'fi; '
@@ -312,20 +315,21 @@ chmod +x /usr/local/bin/rpztar
             # Untar
             if use_chroot:
                 script.write('\n'
-                             'mkdir /experimentroot; cd /experimentroot\n')
+                             'mkdir /experimentroot\n'
+                             'mkdir /.experimentdata; cd /.experimentdata\n')
                 script.write('tar zpxf /vagrant/data.tgz --numeric-owner '
                              '--strip=1 %s\n' % rpz_pack.data_prefix)
                 if mount_bind:
                     script.write('\n'
-                                 'mkdir -p /experimentroot/dev\n'
-                                 'mkdir -p /experimentroot/proc\n')
+                                 'mkdir -p /.experimentdata/dev\n'
+                                 'mkdir -p /.experimentdata/proc\n')
 
                 for pkg in packages:
                     script.write('\n# Copies files from package %s\n'
                                  % pkg.name)
                     for f in pkg.files:
                         f = f.path
-                        dest = join_root(PosixPath('/experimentroot'), f)
+                        dest = join_root(PosixPath('/.experimentdata'), f)
                         script.write('mkdir -p %s\n' %
                                      shell_escape(unicode_(f.parent)))
                         script.write('cp -L %s %s\n' % (
@@ -333,7 +337,7 @@ chmod +x /usr/local/bin/rpztar
                                      shell_escape(unicode_(dest))))
                 script.write(
                     '\n'
-                    'cp /etc/resolv.conf /experimentroot/etc/resolv.conf\n')
+                    'cp /etc/resolv.conf /.experimentdata/etc/resolv.conf\n')
             else:
                 script.write('\ncd /\n')
                 paths = set()
@@ -373,11 +377,11 @@ chmod +x /usr/local/bin/rpztar
                               target / 'busybox',
                               'busybox-%s' % arch)
                 script.write(r'''
-cp /vagrant/busybox /experimentroot/busybox
-chmod +x /experimentroot/busybox
-mkdir -p /experimentroot/bin
-[ -e /experimentroot/bin/sh ] || \
-    ln -s /busybox /experimentroot/bin/sh
+cp /vagrant/busybox /.experimentdata/busybox
+chmod +x /.experimentdata/busybox
+mkdir -p /.experimentdata/bin
+[ -e /.experimentdata/bin/sh ] || \
+    ln -s /busybox /.experimentdata/bin/sh
 ''')
 
         # Copies pack

